From b35b23f74ca32f28ff3a9481efee9c4f0467a759 Mon Sep 17 00:00:00 2001
From: Patrick Hunt <phunt@cloudera.com>
Date: Wed, 23 Jan 2013 11:00:38 -0800
Subject: [PATCH 11/19] CDH-9900. add a configuration option to turn off sending DIFFs to followers

ZooKeeper 3.4 introduced more aggressive use of DIFFs for reconnecting
followers. This improved performace for reconnecting followers after a
leader change, however it resulted in some corruption issues being
exposed. These corruption issues are still being addressed, as such
this patch introduces an option to disable the "aggressive diffing"
feature (now configurable and off by default).
---
 .../server/persistence/FileTxnSnapLog.java         |   11 ++++++++++-
 .../zookeeper/test/RestoreCommittedLogTest.java    |    2 ++
 2 files changed, 12 insertions(+), 1 deletions(-)

diff --git a/src/java/main/org/apache/zookeeper/server/persistence/FileTxnSnapLog.java b/src/java/main/org/apache/zookeeper/server/persistence/FileTxnSnapLog.java
index e80d776..97d32d8 100644
--- a/src/java/main/org/apache/zookeeper/server/persistence/FileTxnSnapLog.java
+++ b/src/java/main/org/apache/zookeeper/server/persistence/FileTxnSnapLog.java
@@ -63,6 +63,11 @@ public class FileTxnSnapLog {
 
     public static final String ZOOKEEPER_DATADIR_AUTOCREATE_DEFAULT = "true";
 
+    public static final String ZOOKEEPER_CDH_RESTORE_COMMITTED_LOG =
+            "zookeeper.cdh.restoreCommittedLog";
+    private static final boolean RESTORE_COMMITTED_LOG =
+            Boolean.getBoolean(ZOOKEEPER_CDH_RESTORE_COMMITTED_LOG);
+
     /**
      * This listener helps
      * the external apis calling
@@ -183,7 +188,11 @@ public class FileTxnSnapLog {
                throw new IOException("Failed to process transaction type: " +
                      hdr.getType() + " error: " + e.getMessage(), e);
             }
-            listener.onTxnLoaded(hdr, itr.getTxn());
+            /* Addresses CDH-9900 - reduce the likelihood of sending a diff
+             * on leader change */
+            if (RESTORE_COMMITTED_LOG) {
+                listener.onTxnLoaded(hdr, itr.getTxn());
+            }
             if (!itr.next()) 
                 break;
         }
diff --git a/src/java/test/org/apache/zookeeper/test/RestoreCommittedLogTest.java b/src/java/test/org/apache/zookeeper/test/RestoreCommittedLogTest.java
index 9a9ff83..b06ce53 100644
--- a/src/java/test/org/apache/zookeeper/test/RestoreCommittedLogTest.java
+++ b/src/java/test/org/apache/zookeeper/test/RestoreCommittedLogTest.java
@@ -51,6 +51,8 @@ public class RestoreCommittedLogTest extends ZKTestCase implements  Watcher {
      */
     @Test
     public void testRestoreCommittedLog() throws Exception {
+        System.setProperty(FileTxnSnapLog.ZOOKEEPER_CDH_RESTORE_COMMITTED_LOG, "true");
+
         File tmpDir = ClientBase.createTmpDir();
         ClientBase.setupTestEnv();
         ZooKeeperServer zks = new ZooKeeperServer(tmpDir, tmpDir, 3000);
-- 
1.7.0.4

